src/lib.rs

```diff
@@ -2722,3 +2710,64 @@ fn command_add_output_file(
         cmd.arg("-o").arg(&dst);
     }
 }
+
+static NEW_STANDALONE_ANDROID_COMPILERS: [&str; 4] = [
+    "aarch64-linux-android21-clang",
+    "armv7a-linux-androideabi16-clang",
+    "i686-linux-android16-clang",
+    "x86_64-linux-android21-clang",
+];
+
+fn new_clang_android_compiler(clang_path: &Path) -> bool {
+    NEW_STANDALONE_ANDROID_COMPILERS.iter().any(|x| {
+        let x: &OsStr = x.as_ref();
+        x == clang_path.as_os_str()
+    })
+}
+
+fn autodetect_android_compiler(target: &str, host: &str, gnu: &str, clang: &str) -> String {
+    let new_clang_key = match target {
+        "aarch64-linux-android" => Some("aarch64"),
+        "armv7-linux-androideabi" => Some("armv7a"),
+        "i686-linux-android" => Some("i686"),
+        "x86_64-linux-android" => Some("x86_64"),
+        _ => None,
+    };
+
+    let new_clang = new_clang_key
+        .map(|key| {
+            NEW_STANDALONE_ANDROID_COMPILERS
+                .iter()
+                .find(|x| x.starts_with(key))
+        })
+        .unwrap_or(None);
+
+    if let Some(new_clang) = new_clang {
+        if Command::new(new_clang).output().is_ok() {
+            return (*new_clang).into();
+        }
+    }
+
+    let target = target
+        .replace("armv7neon", "arm")
+        .replace("armv7", "arm")
+        .replace("thumbv7neon", "arm")
+        .replace("thumbv7", "arm");
```

Comment:
FWIW this normalization isn't used when trying to find a new compiler, could this be refactored a bit such that these rustc targets would still find the armv7a android compiler?

Reply:
I intentionally did not touch this code. Tier 1,2 platforms have only linux+android support. So it is hard test for Windows and MacOS, but code bellow works with windows, so I avoid changing anything that I can not test. 

Plus I don't think that this is refactoring is possible. IMHO new target naming that include android API level was introduced recently, so if somebody do not want update Android NDK and uses old `clang` it is easy enough to break things for them.

Reply:
My point is that the `armv7-linux-androideabi` target, if folks are using it, will not hit the new clang compiler when it very likely should. I don't think there's a lot of risk here, it's basically just moving this `let target` block up to the top of the function instead of having it at the bottom

Reply:
> My point is that the armv7-linux-androideabi target, if folks are using it, will not hit the new 
> clang compiler when it very likely should.

@alexcrichton, not sure that I understand you. `armv7-linux-androideabi` mapped to Some("armv7a") and then to `armv7a-linux-androideabi16-clang` call.
So `armv7-linux-androideabi` hit the new clang compiler, it is my main testing configuration,
I build all my code with `cargo build --target=armv7-linux-androideabi` and run on android device before submitting this PR.

Reply:
Ah ok I think I was misreading a bit. In any case we can always tweak this later!